capirca/lib/versa.py

```diff
@@ -0,0 +1,877 @@
+      self.BuildTermLogging(access_pn)
+
+    #print("\n".join(set_term.PrintTree()))
+
+
+
+class Versa(aclgenerator.ACLGenerator):
+  """Versa rendering class.
+
+     This class takes a policy object and renders the output into a syntax
+     which is understood by Versa firewalls.
+
+     Args:
+       pol: policy.Policy object
+  """
+
+  _PLATFORM = 'versa'
+  SUFFIX = '.vsp'
+  _SUPPORTED_AF = set(('inet', 'inet6', 'mixed'))
+  _ZONE_ADDR_BOOK = 'address-book-zone'
+  _GLOBAL_ADDR_BOOK = 'address-book-global'
+  _ADDRESSBOOK_TYPES = set((_ZONE_ADDR_BOOK, _GLOBAL_ADDR_BOOK))
+  _NOVERBOSE = 'noverbose'
+  _SUPPORTED_TARGET_OPTIONS = set((_ZONE_ADDR_BOOK,
+                                   _GLOBAL_ADDR_BOOK,
+                                   _NOVERBOSE))
+  _VERSA_SUPPORTED_TARGET_OPTIONS = set(('template',
+                                         'tenant',
+                                         'policy'))
+  _VERSA_UNSUPPORTED_TERM_OPTIONS = set(('icmp_type',
+                                         'stateless_reply',
+                                         'translated',
+                                         'verbatim'))
+
+  _AF_MAP = {'inet': (4,),
+             'inet6': (6,),
+             'mixed': (4, 6)}
+  _AF_ICMP_MAP = {'icmp': 'inet',
+                  'icmpv6': 'inet6'}
+  INDENT = '    '
+  _MAX_HEADER_COMMENT_LENGTH = 71
+  # The Versa platform is limited in how many IP addresses can be used in
+  # a single policy.
+  _ADDRESS_LENGTH_LIMIT = 1023
+  # IPv6 are 32 bytes compared to IPv4, this is used as a multiplier.
+  _IPV6_SIZE = 4
+
+  def __init__(self, pol, exp_info):
+    self.versa_policies = []
+    self.comment = ''
+    self.addressbook = collections.OrderedDict()
+    self.applications = []
+    self.ports = []
+    self.from_zone = ''
+    self.to_zone = ''
+    self.addr_book_type = set()
+    self.templatename = '_templatename'
+    self.tenantname = '_tenantname'
+    self.policyname = '_policyname'
+    super().__init__(pol, exp_info)
+
+  def _BuildTokens(self):
+    """Build supported tokens for platform.
+
+    Returns:
+      tuple containing both supported tokens and sub tokens
+    """
+    supported_tokens, supported_sub_tokens = super()._BuildTokens()
+
+    supported_tokens |= {
+                         'dscp_match',
+                         'destination_zone',
+                         'logging',
+                         'option',
+                         'source_zone',
+                         'versa_application'
+                         }
+
+    supported_sub_tokens.update(
+        {'action': {'accept', 'deny', 'reject', 'count', 'log', 'dscp'},
+        })
+
+    del supported_sub_tokens['option']
+    return supported_tokens, supported_sub_tokens
+
+
+  def HeaderParams(self, mstr, val):
+    """HeaderParams populates the template name and tenant name."""
+    if len(val) > 0:
+      if 'template' in mstr:
+        self.templatename = val
+      elif 'tenant' in mstr:
```

Comment:
Is tenant a required field? Could you provide an example of what might be populated in here?

Reply:
Added a skeleton of the config.